### PR TITLE
Detect duplicates early

### PR DESCRIPTION
**Pull request by mdiep**
This is the first step in some cleanup I want to do—changing Cartfiles to have `Set`s of dependencies instead of `Array`s, changing `Resolver` to not know about `Cartfile`s, etc.

### CODE REVIEW

File: Source/CarthageKitTests/CartfileSpec.swift [hunk 1]
```diff
@@ -74,23 +74,26 @@ class CartfileSpec: QuickSpec {
...
+					.sort { $0.description < $1.description }
+				expect(dupes.count) == 2
+				
+				let self2Dupe = dupes[0]
+				expect(self2Dupe) == ProjectIdentifier.gitHub(Repository(owner: "self2", name: "self2"))
+				
+				let self3Dupe = dupes[1]
+				expect(self3Dupe) == ProjectIdentifier.gitHub(Repository(owner: "self3", name: "self3"))
+			} else {
+				fatalError("should error")
```
Comment (mdiep) [inline Source/CarthageKitTests/CartfileSpec.swift]: I wasn't sure if there was a better way to do this with Quick/Nimble. If we were using XCTest directly, I'd use `XCFail`. But this certainly works. 😕

Comment (ikesyo) [inline Source/CarthageKitTests/CartfileSpec.swift]: FYI Nimble has `fail` functions: https://github.com/Quick/Nimble/blob/v5.1.1/Sources/Nimble/DSL.swift#L21-L35.

Comment (mdiep) [inline Source/CarthageKitTests/CartfileSpec.swift]: Thanks! Xcode's code completion wasn't working. 🙃

**Comment by ikesyo** [comment]
`let mainLocation = ...` in `ProjectSpec.swift` could be removed.

File: Source/CarthageKit/Cartfile.swift [hunk 2]
```diff
@@ -69,14 +69,40 @@ public struct Cartfile {
...
 
 	/// Attempts to parse a Cartfile from a file at a given URL.
 	public static func from(file cartfileURL: URL) -> Result<Cartfile, CarthageError> {
 		do {
 			let cartfileContents = try String(contentsOf: cartfileURL, encoding: .utf8)
-			return Cartfile.from(string: cartfileContents)
+			return Cartfile
+				.from(string: cartfileContents)
+				.mapError { error in
+					if case let .duplicateDependencies(dupes) = error {
```
Comment (ikesyo) [inline Source/CarthageKit/Cartfile.swift]: `guard` would be preferable:

```swift
guard case let .duplicateDependencies(dupes) = error else {
	return error
}

let dupes = dupes
	.map { dependency in
		return DuplicateDependency(
			project: dependency.project,
			locations: [ cartfileURL.carthage_path ]
		)
	}
return .duplicateDependencies(dupes)
```

File: Source/CarthageKitTests/CartfileSpec.swift [hunk 3]
```diff
@@ -74,23 +74,26 @@ class CartfileSpec: QuickSpec {
...
-			expect(dupes.count) == 2
-
-			let self2Dupe = dupes[0]
-			expect(self2Dupe) == ProjectIdentifier.gitHub(Repository(owner: "self2", name: "self2"))
-
-			let self3Dupe = dupes[1]
-			expect(self3Dupe) == ProjectIdentifier.gitHub(Repository(owner: "self3", name: "self3"))
+			expect(result.error).notTo(beNil())
+
+			if case let .duplicateDependencies(dupes)? = result.error {
```
Comment (ikesyo) [inline Source/CarthageKitTests/CartfileSpec.swift]: Ditto:

```swift
guard case let .duplicateDependencies(dupes)? = result.error else {
	fail("Cartfile should error with duplicate dependencies")
}

let dupes = dupes
	.map { $0.project }
	.sort { $0.description < $1.description }
expect(dupes.count) == 2

let self2Dupe = dupes[0]
expect(self2Dupe) == ProjectIdentifier.gitHub(Repository(owner: "self2", name: "self2"))

let self3Dupe = dupes[1]
expect(self3Dupe) == ProjectIdentifier.gitHub(Repository(owner: "self3", name: "self3"))
```

**Pull request merged**